### PR TITLE
Implement binary_search for `collections::Vec`

### PR DESCRIPTION
**Pull request by KaiserKarel**
This ports the binary_search implementation from `core`. I've removed the use of `get_unchecked` as that is not present in `LazyIndexMap`.  There's 2 tests added to verify that it works with a full vector and an empty vector. 

related to #834

### CODE REVIEW

**Comment by KaiserKarel** [comment]
I think that the spellcheck failure is a false alarm, as the documentation is taken from core.

File: crates/storage/src/collections/vec/mod.rs [hunk 1]
```diff
@@ -200,6 +200,169 @@ where
...
+        let mut size = self.len();
+        let mut left = 0;
+        let mut right = size;
+        while left < right {
+            let mid = left + size / 2;
+
+            // SAFETY: the call is made safe by the following invariants:
+            // - `mid >= 0`
+            // - `mid < size`: `mid` is limited by `[left; right)` bound.
+            let cmp = f(self.elems.get(mid).unwrap());
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:305]:
```suggestion
            let cmp = f(&self.elems[mid]);
```

File: crates/storage/src/collections/vec/mod.rs [hunk 2]
```diff
@@ -200,6 +200,169 @@ where
...
+    {
+        use core::cmp::Ordering::*;
+
+        let mut size = self.len();
+        let mut left = 0;
+        let mut right = size;
+        while left < right {
+            let mid = left + size / 2;
+
+            // SAFETY: the call is made safe by the following invariants:
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:302]: `// SAFETY` comments are only needed for `unsafe` code blocks, although I do appreciate the comment in general here.

File: crates/storage/src/collections/vec/mod.rs (hunk 1 above)
Comment (Robbepop) [reply crates/storage/src/collections/vec/mod.rs:305]: Both are doing the same, the one is just shorter and a bit more idiomatic in this case of immediate `unwrap`.

**Review by Robbepop — COMMENTED** [review]
All in all LGTM so far. Will take another more detailed look into the actual implementation. We could use a bit more tests for this to cover some edge cases. Maybe pull some tests over from the `core` implementation?

File: crates/storage/src/collections/vec/mod.rs (hunk 1 above)
Comment (KaiserKarel) [reply crates/storage/src/collections/vec/mod.rs:305]: agree!

File: crates/storage/src/collections/vec/mod.rs (hunk 2 above)
Comment (KaiserKarel) [reply crates/storage/src/collections/vec/mod.rs:302]: That's a leftover of the actual vec implementation having a get_unchecked method for perf reasons. Perhaps I should remove the `SAFETY:`, but leave the comment?

**Comment by KaiserKarel** [comment]
> All in all LGTM so far. Will take another more detailed look into the actual implementation. We could use a bit more tests for this to cover some edge cases. Maybe pull some tests over from the `core` implementation?

I originally was unable to find them, but now realize they resided in a different crate. I've added the core tests as well now, and removed my own tests.

**Comment by KaiserKarel** [comment]
Also @Robbepop, sinc binary_search and partition are quite closely related, should I add that one too? Or perhaps create another issue and another PR?

**Comment by Robbepop** [comment]
> Also @Robbepop, sinc binary_search and partition are quite closely related, should I add that one too? Or perhaps create another issue and another PR?

Another API should be done in another PR. :)

**Review by Robbepop — COMMENTED** [review]
Doc comments for the functions need to be adjusted to fit ink! instead of Rust standard library. It currently contains dead links and invalid docs or docs that just does not fit. Please rework docs in this regard and we are good to go. Also we might be in need of a `sort` function for `storage::Vec` in the near future. Although sorting efficiently will be a major endavour and might probably be a bad idea.

**Comment by KaiserKarel** [comment]
> Doc comments for the functions need to be adjusted to fit ink! instead of Rust standard library. It currently contains dead links and invalid docs or docs that just does not fit. Please rework docs in this regard and we are good to go. Also we might be in need of a `sort` function for `storage::Vec` in the near future. Although sorting efficiently will be a major endavour and might probably be a bad idea.

For sorting it would probably be for efficient to read the storage vec into a normal vec, sort that, and write the result back? That should minimize the amount of gas fees I think.

**Comment by Robbepop** [comment]
> For sorting it would probably be for efficient to read the storage vec into a normal vec, sort that, and write the result back? That should minimize the amount of gas fees I think.

Maybe yes, however due to caching in the `LazyIndexMap` layer this might not be needed. Benchmarks are required here instead of assumptions about performance.

**Comment by KaiserKarel** [comment]
Alright, so a PR involving adding sorting would require benchmarks of the two competing implementations?

**Review by Robbepop — APPROVED** [review]
LGTM
@cmichi please review/approve too.

We need to check the spellcheck CI so that our full pipeline is run on this PR before we merge.

File: crates/storage/src/collections/vec/tests.rs [hunk 3]
```diff
@@ -422,11 +423,88 @@ fn clear_works_on_filled_vec() {
...
+        b.binary_search_by(|_| Ordering::Equal),
+        Ok(u8::MAX as u32 / 2)
+    );
+    assert_eq!(b.binary_search_by(|_| Ordering::Greater), Err(0));
+    assert_eq!(b.binary_search_by(|_| Ordering::Less), Err(u8::MAX as u32));
+}
+
+#[test]
+// Test implementation specific behavior when finding equivalent elements.
+// It is ok to break this test but when you do a crater run is highly advisable.
```
Comment (HCastano) [inline crates/storage/src/collections/vec/tests.rs:494]: This comment references `crater`, which isn't relevant in our case. We should probably remove that part of the comment.

File: crates/storage/src/collections/vec/mod.rs [hunk 4]
```diff
@@ -200,6 +200,163 @@ where
         *self.len += 1;
         self.elems.put(last_index, Some(value));
     }
+
+    /// Binary searches this sorted vector for a given element.
```
Comment (HCastano) [inline crates/storage/src/collections/vec/mod.rs:204]: Can you add a comment explaining where this comes from and a link to the original source code? It doesn't need to be part of the doc comments since it would mostly be of interest to devs working on the internals

**Review by HCastano — APPROVED** [review]

File: crates/storage/src/collections/vec/mod.rs [hunk 5]
```diff
@@ -200,6 +200,165 @@ where
...
+    /// let seek = 4;
+    /// assert_eq!(s.binary_search_by(|probe| probe.cmp(&seek)), Err(7));
+    /// let seek = 100;
+    /// assert_eq!(s.binary_search_by(|probe| probe.cmp(&seek)), Err(13));
+    /// let seek = 1;
+    /// let r = s.binary_search_by(|probe| probe.cmp(&seek));
+    /// assert!(match r { Ok(1..=4) => true, _ => false, });
+    /// ```
+    // The binary_search implementation is ported from [core::slice] and attempts to remain as close
+    // to the source as possible.
```
Comment (HCastano) [inline crates/storage/src/collections/vec/mod.rs:289]: Do you mind adding a URL to the source code? One of those "permanent" GitHub URLs would be best

File: crates/storage/src/collections/vec/tests.rs (hunk 3 above)
Comment (HCastano) [reply crates/storage/src/collections/vec/tests.rs]: We should leave the first part of the comment ("it is ok to break...") here, it's just the `crater` part that's not relevant

**Review by HCastano — COMMENTED** [review]
Some small nit picks, but after they're resolved we can merge

**Comment by KaiserKarel** [comment]
> We should leave the first part of the comment ("it is ok to break...") here, it's just the crater part that's not relevant

Actually now that I think about it, wasn't binary_search implementation specific behaviour the reason that polkadot had an outage? Perhaps it is not okay for this test to break, as contracts may rely on the specific behaviour?

**Comment by HCastano** [comment]
> Actually now that I think about it, wasn't binary_search implementation specific behaviour the reason that polkadot had an outage? Perhaps it is not okay for this test to break, as contracts may rely on the specific behaviour?

Umm kinda. The problem with Polkadot was that the Wasm runtime and native runtime had different sorting behaviours which caused a failure in consensus. You can read more about it [here](https://polkadot.network/a-polkadot-postmortem-24-05-2021/). 

In our case once a contract is deployed that's the only instance of it. There is no need for a network of people with potentially different versions of the contract to come to consensus on the behaviour - they only need to run what's already on chain.

Whether or not we want to make any guarantees around sorting behaviour is an interesting topic. I'm not sure what the answer is here though. @cmichi @Robbepop what do you guys think?

**Comment by Robbepop** [comment]
> Whether or not we want to make any guarantees around sorting behaviour is an interesting topic. I'm not sure what the answer is here though. @cmichi @Robbepop what do you guys think?

We should do the same as Rust and provide no sorting behavior since this could in theory severely reduce the efficiency of the algorithm. Furthermore, the API clearly says this and the problem for Polkadot existed because Polkadot did not obey the API calling guarantees and took something for granted that simply did not exist. So the problem was not caused by the Rust API or implementation but purely on the user side (Polkadot).

File: crates/storage/src/collections/vec/mod.rs [hunk 6]
```diff
@@ -200,6 +200,166 @@ where
...
+    ///
+    /// If you want to insert an item to a sorted vector, while maintaining
+    /// sort order:
+    ///
+    /// ```
+    /// let mut s = Vec::from([0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55]);
+    /// let num = 42;
+    /// let idx = s.binary_search(&num).unwrap_or_else(|x| x);
+    /// s.insert(idx, num);
+    /// assert_eq!(s, [0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 42, 55]);
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:241]: This example should ideally show `ink_storage::Vec` instead of Rust's `Vec`.

File: crates/storage/src/collections/vec/mod.rs [hunk 7]
```diff
@@ -200,6 +200,166 @@ where
...
+    /// found; the fourth could match any position in `[1, 4]`.
+    ///
+    /// ```
+    /// let s = Vec::from([0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55]);
+    ///
+    /// assert_eq!(s.binary_search(&13),  Ok(9));
+    /// assert_eq!(s.binary_search(&4),   Err(7));
+    /// assert_eq!(s.binary_search(&100), Err(13));
+    /// let r = s.binary_search(&1);
+    /// assert!(match r { Ok(1..=4) => true, _ => false, });
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:230]: This example should ideally show ink_storage::Vec instead of Rust's Vec.

File: crates/storage/src/collections/vec/mod.rs [hunk 8]
```diff
@@ -200,6 +200,166 @@ where
...
+    ///
+    /// let seek = 13;
+    /// assert_eq!(s.binary_search_by(|probe| probe.cmp(&seek)), Ok(9));
+    /// let seek = 4;
+    /// assert_eq!(s.binary_search_by(|probe| probe.cmp(&seek)), Err(7));
+    /// let seek = 100;
+    /// assert_eq!(s.binary_search_by(|probe| probe.cmp(&seek)), Err(13));
+    /// let seek = 1;
+    /// let r = s.binary_search_by(|probe| probe.cmp(&seek));
+    /// assert!(match r { Ok(1..=4) => true, _ => false, });
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:286]: This example should ideally show ink_storage::Vec instead of Rust's Vec.

File: crates/storage/src/collections/vec/mod.rs [hunk 9]
```diff
@@ -200,6 +200,166 @@ where
...
+    /// ```
+    /// let s = Vec::from([(0, 0), (2, 1), (4, 1), (5, 1), (3, 1),
+    ///          (1, 2), (2, 3), (4, 5), (5, 8), (3, 13),
+    ///          (1, 21), (2, 34), (4, 55)]);
+    ///
+    /// assert_eq!(s.binary_search_by_key(&13, |&(a, b)| b),  Ok(9));
+    /// assert_eq!(s.binary_search_by_key(&4, |&(a, b)| b),   Err(7));
+    /// assert_eq!(s.binary_search_by_key(&100, |&(a, b)| b), Err(13));
+    /// let r = s.binary_search_by_key(&1, |&(a, b)| b);
+    /// assert!(match r { Ok(1..=4) => true, _ => false, });
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:353]: This example should ideally show ink_storage::Vec instead of Rust's Vec.

**Review by Robbepop — COMMENTED** [review]
Please fix the doc examples to work with the respecitve `ink_storage::Vec` type so that we can merge this.

File: crates/storage/src/collections/vec/mod.rs (hunk 9 above)
Comment (KaiserKarel) [reply crates/storage/src/collections/vec/mod.rs:353]: Oof you're right, that's missing an import. Do I rename the import to StorageVec, to reuse the same nomenclature as what is generated by the templates?
